src/lib.rs

```diff
@@ -115,43 +126,199 @@ impl Config {
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// let io = r#"
+    /// [[redirects]]
+    ///   from = "/api/*"
+    ///   to = "https://production.api.com/:splat"
+    ///
+    /// [[context.deploy-preview.redirects]]
+    ///   from = "/api/*"
+    ///   to = "https://staging.api.com/:splat"
+    /// "#;
+    ///
+    /// let config = netlify_toml::from_str(io).unwrap();
+    /// let redirects = config.scoped_redirects("deploy-preview", "new-styles").expect("missing redirects");
+    /// assert_eq!(1, redirects.len());
+    /// let dest = redirects.first().and_then(|r| r.to.as_ref()).expect("missing destination");
+    /// assert_eq!("https://staging.api.com/:splat", dest.as_str());
+    /// ```
+    pub fn scoped_redirects(&self, ctx: &str, branch: &str) -> Option<Vec<Redirect>> {
+        let context = match &self.context {
+            Some(c) => c,
+            None => return self.redirects.clone(),
+        };
+
+        let mut global_index = IndexMap::new();
+        if let Some(global) = &self.redirects {
+            for r in global {
+                global_index.insert(r.from.clone(), r.clone());
             }
         }
 
-        if let Some(c) = self.context {
-            // Override with default context environment,
-            // like `deploy-preview`, `branch-deploy` or `production`.
-            if let Some(x) = c.get(ctx) {
-                if let Some(ref env) = x.environment {
-                    for (k, v) in env {
-                        result.insert(k.to_string(), v.to_string());
-                    }
+        if let Some(ct) = context.get(ctx).and_then(|x| x.redirects.as_ref()) {
+            for r in ct {
+                if global_index.contains_key(&r.from) {
+                    global_index.insert(r.from.clone(), r.clone());
+                } else {
+                    let mut swap = IndexMap::<String, Redirect>::new();
+                    swap.insert(r.from.clone(), r.clone());
+                    swap.extend(global_index);
+                    global_index = swap;
                 }
             }
+        };
 
-            // Override with branch context environment,
-            // like `deploy-preview`, `branch-deploy` or `production`.
-            if let Some(x) = c.get(branch) {
-                if let Some(ref env) = x.environment {
-                    for (k, v) in env {
-                        result.insert(k.to_string(), v.to_string());
-                    }
+        if let Some(ct) = context.get(branch).and_then(|x| x.redirects.as_ref()) {
+            for r in ct {
+                if global_index.contains_key(&r.from) {
+                    global_index.insert(r.from.clone(), r.clone());
+                } else {
+                    let mut swap = IndexMap::<String, Redirect>::new();
+                    swap.insert(r.from.clone(), r.clone());
+                    swap.extend(global_index);
+                    global_index = swap;
                 }
             }
+        };
```

Comment:
this makes the order of the prepended rules a bit more obvious and only runs the swap once.

```suggestion
        let mut insert = Vec::new();

        if let Some(ct) = context.get(ctx).and_then(|x| x.redirects.clone()) {
            for r in ct {
                if global_index.contains_key(&r.from) {
                    global_index.insert(r.from.clone(), r);
                } else {
                    insert.push(r);
                }
            }
        };

        if let Some(ct) = context.get(branch).and_then(|x| x.redirects.clone()) {
            for r in ct {
                if global_index.contains_key(&r.from) {
                    global_index.insert(r.from.clone(), r);
                } else {
                    insert.push(r);
                }
            }
        };

        if !insert.is_empty() {
            let mut swap = IndexMap::<String, Redirect>::new();
            swap.extend(insert.into_iter().map(|r| (r.from.clone(), r)));
            swap.extend(global_index);
            global_index = swap;
        }
```